tests/unit/db/query_result_tests.py

```diff
@@ -0,0 +1,380 @@
+
+    def test_iteration_with_skip_option_fails(self):
+        """
+        Test __iter__() fails when "skip" option is provided
+        """
+        self.populate_db_with_documents(100)
+        result = QueryResult(
+            Query(self.db),
+            selector={'_id': {'$gt': 0}},
+            fields=['_id', '_rev'],
+            skip=10
+        )
+        try:
+            for doc in result:
+                self.fail('Above statement should raise an Exception')
+        except CloudantArgumentError, err:
+            self.assertEqual(
+                str(err),
+                'Cannot use skip for iteration'
+            )
+
+    def test_iteration_with_limit_option_fails(self):
+        """
+        Test __iter__() fails when "limit" option is provided
+        """
+        self.populate_db_with_documents(100)
+        result = QueryResult(
+            Query(self.db),
+            selector={'_id': {'$gt': 0}},
+            fields=['_id', '_rev'],
+            limit=10
+        )
+        try:
+            for doc in result:
+                self.fail('Above statement should raise an Exception')
+        except CloudantArgumentError, err:
+            self.assertEqual(
+                str(err),
+                'Cannot use limit for iteration'
+            )
+
+    def test_iteration_invalid_page_size(self):
+        """
+        Test __iter__() fails as expected when page_size is set to 0 or less
+        """
+        self.populate_db_with_documents(5)
+        result = QueryResult(
+            Query(self.db),
+            selector={'_id': {'$gt': 0}},
+            fields=['_id', '_rev'],
+            page_size=0
+        )
+        try:
+            for doc in result:
+                self.fail('Above statement should raise an Exception')
+        except CloudantArgumentError, err:
+            self.assertEqual(str(err), 'Invalid page_size: 0')
+
+    def test_iteration_result_eq_page_size(self):
+        """
+        Test __iter__() works as expected when the result size is equal to the
+        page_size
+        """
+        self.populate_db_with_documents(5)
+        result = QueryResult(
+            Query(self.db),
+            selector={'_id': {'$gt': 0}},
+            fields=['_id', '_rev'],
+            page_size=5
+        )
+        self.assertEqual(
+            [doc['_id'] for doc in result],
+            ['julia000', 'julia001', 'julia002', 'julia003', 'julia004']
+        )
+
+    def test_iteration_result_gt_page_size(self):
+        """
+        Test __iter__() works as expected when the result size is more than the
+        page_size
+        """
+        self.populate_db_with_documents(5)
+        result = QueryResult(
+            Query(self.db),
+            selector={'_id': {'$gt': 0}},
+            fields=['_id', '_rev'],
+            page_size=3
+        )
+        self.assertEqual(
+            [doc['_id'] for doc in result],
+            ['julia000', 'julia001', 'julia002', 'julia003', 'julia004']
```

Comment:
There appear to be 5 results here even though the page size is 3. Is this because the iterator has all the results and the QueryResult is what filters it down to 3? I was a bit confused because I thought skip/limit couldn't be used with the page size because they were used under the covers for the pages, in which case I would not have expected documents exceeding the page size to have been returned from the server.